### ISSUE TITLE
Align with expected nvim-treesitter structure

### ISSUE THREAD

**Issue by gamebox** [faldor20/tree-sitter-roc#26]
I'm happy to put up a pull request for this if needed.

Right how there are two things that stop this from being used easily with Neovim:

1. The lua code is in a `/neovim` file.  It should be `/lua/tree-sitter-roc`
2. The queries need to be in a `/queries/roc` folder

And that's it!

Let me know if you are interested in a PR

**Comment by gamebox** [comment]
And if this is done, I can put a PR up on nvim-treesitter to add Roc as a default language config to download.

That means people would just need to add "roc" to their ensure_installed, and restart their editor.  Well, that and adding the filetype.

**Comment by faldor20** [comment]
Okay, we already had some work in this area: https://github.com/faldor20/tree-sitter-roc/pull/16
As someone who doesn't use neovim I'm kind of just trusting that you folks know what's right. 
@nat-418 Can you comment on this suggestion?

**Comment by gamebox** [comment]
I have it on my fork and was able to install with just adding the install info and then doing a :TsInstall roc_ls.  Worked like a dream

**Comment by faldor20** [comment]
Sounds good! Make a pr and update the readme and I'll merge it :) 
Thanks for offering to help!

**Comment by clason** [comment]
PSA: Someone added this to nvim-treesitter, so there's nothing that needs to be done here.

BUT someone needs to take ownership of the queries and keep them compatible to the (recent _very_) breaking changes here or I will have to drop them again.

**Comment by clason** [comment]
(In fact, the config here is incompatible with the current version of nvim-treesitter, so might as well drop it -- like you did the Neovim queries.)

**Comment by faldor20** [comment]
Hey @clason, we're just in the process of updating all the editors to the latest version.
Would you be able to explain why the queries aren't working with neovim? 
I don't use nvim so I haven't been a able to test. 
My hope was that by combining the queries for all the editors and just changing the capture names I'd have a better time keeping them in sync. 
I'll get neovim setup in the next few days to test with if you're not able to help.

**Comment by clason** [comment]
Well, there's two things here: the "plugin" and the queries. I am specifically talking about the plugin (so the lua/vimscript files), which should just be removed.

Regarding the queries: Those actually live in nvim-treesitter since they need to be editor-specific (and tested!) and consistent across languages; we treat queries in parser repositories merely as reference, especially how to adapt to breaking changes. (It's not just capture names but also predicates and directives; for example, `#match?` in Neovim is extremely slow since it has to go through Vim's regex engine, so we often need to refactor those into Lua patterns which are much faster.)

However, this only works to some extent: if the changes are too big and the queries have diverged too much (as is the case here), this requires more familiarity with the language and the parser than I have. This is where a dedicated "Neovim queries maintainer" is required. Unfortunately, people don't take this seriously enough -- we get a lot of eager "drive-by" contributions of new languages with little or no follow-up, especially long-term.

So TL;DR:
1. just drop the "plugin" part (it's not needed for current Neovim) and focus on grammar and (reference) queries;
2. don't worry about editor-specific queries and focus on keeping the reference queries clean and compatible (use https://github.com/ribru17/ts_query_ls in a workflow);
3. nvim-treesitter needs an active "Roc" maintainer (not you!) who follows grammar development and updates the nvim-treesitter queries to major breaking changes in a timely manner.

**Comment by faldor20** [comment]
Ah, I see. Thank you for the clarification. I really appreciate it!
I'll definitely checkout query-ls.

I think in the short term, I'll do my best to provide some queries that at least work in nvim and documentation for how to add them. 

And in the long term, obviously it would be good to have a maintainer for the main nvim-treesitter queries, like you mentioned. 

There are definitely bugs in the current grammar and queries that will be resolved in the next couple of weeks while some real people are actually using it. Hopefully by the time it gets added to nvim-treesitter those will all be sorted.

**Comment by faldor20** [comment]
I must say, the whole editor-specific queries thing does feel like a real weakness of treesitter and it's quite frustrating that there isn't some canonical spec that editors at least try to conform to.

**Comment by clason** [comment]
> Hopefully by the time it gets added to nvim-treesitter those will all be sorted.

I have "locked" the parser and queries, so nothing will (or needs to) happen in the near future. It's better to sort things out fully and _then_ (wait for someone to) make a PR to nvim-treesitter.

In the mid-term, look into making (semver) releases using the upstream workflows; then we can switch to tracking releases instead of commits in nvim-treesitter, which means updates will only (need to) happen when _you_ decide.

> I must say, the whole editor-specific queries thing does feel like a real weakness of treesitter and it's quite frustrating that there isn't some canonical spec that editors at least try to conform to.

I felt similarly, but then I realized that queries are the "glue" between tree-sitter parsers and the editor -- and so it makes total sense for them to be editor-specific. (You can no more have a canonical spec for queries than you can have for editors...)

I think the "reference queries you can (hopefully easily) customize" is the correct mental model and comes with reasonable maintenance load for parser developers.
